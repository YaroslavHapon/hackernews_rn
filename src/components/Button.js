import React from 'react'
import { View, Text, TouchableOpacity  } from 'react-native'
import PropTypes from 'prop-types'

const Button = ({ text, onPress, style, buttonTextStyle}) => {
  return(
    <TouchableOpacity onPress={onPress}>
      <View style={style}>
        <Text style={buttonTextStyle}>{text}</Text>
      </View>
    </TouchableOpacity>
  )
};
Button.propTypes = {
  text: PropTypes.string,
  onPress: PropTypes.func,
  style: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.number
  ]),
  buttonTextStyle: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.number
  ])
}
export default Button