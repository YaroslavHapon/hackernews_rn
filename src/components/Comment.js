import React from 'react'
import { View, Text, StyleSheet, ActivityIndicator, Linking } from 'react-native'
import HTMLView from 'react-native-htmlview';
import CommentContainer from './../containers/CommentContainer'
import Button from './Button'
import PropTypes from 'prop-types'

const styles = StyleSheet.create({
  commentContainer: {
    flex: 1,
    padding: 10,
    backgroundColor: "#fff"
  },
  loadingContainer: {
    flex: 1,
    paddingTop: 20
  },
  kidsContainer: {
    borderLeftWidth: 1,
    borderColor: '#888',
    marginLeft: 10,
    marginTop: 10,
    paddingLeft: 10
  },
  author: {
    fontWeight: 'bold'
  },
  button: {
    alignItems: 'flex-end',
    marginTop: 10,
    marginBottom: 10
  },
  buttonTextStyle: {
    color: '#333',
    padding: 5,
    fontWeight: 'bold',
    fontSize: 12
  }
});

export default class Comment extends React.Component{
  static propTypes = {
    fetchComment: PropTypes.func,
    commentGet: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.shape({
        loading: PropTypes.bool,
        loaded: PropTypes.bool,
        data: PropTypes.object,
        error: PropTypes.object
      })
    ]),
    id: PropTypes.number
  }

  state = {
    showAnswers: false
  }

  componentWillMount () {
    this.props.fetchComment(this.props.id);
  }

  renderKidsComments = kids => {
    return kids.map((kid, index) => <CommentContainer innerStyle={styles.kidsContainer} key={index} id={kid} />);
  }

  render () {
    const { commentGet } = this.props

    if(!commentGet[this.props.id]) {
      return <View style={styles.loadingContainer} />;
    }
    
    if (commentGet[this.props.id].loading) {
      return (
        <View style={styles.loadingContainer}>
          <ActivityIndicator />
        </View>
      );
    }

    if (commentGet[this.props.id].error) {
      return (
        <View>{commentGet[this.props.id].error.message}</View>
      );
    }

    const html = commentGet[this.props.id].data.text;
    const kids = commentGet[this.props.id].data.kids;

    return(
      <View style={this.props.innerStyle || styles.commentContainer}>
        <Text style={styles.author}>Author: {commentGet[this.props.id].data.by}</Text>
        <HTMLView
          style={styles.webView}
          onLinkPress={(url) => Linking.openURL(url)}
          value={html} />
        {kids &&
          <View>
            <Button
              text={!this.state.showAnswers ? 'Show Answers ↓' : 'Hide Answers ↑'}
              style={styles.button}
              buttonTextStyle={styles.buttonTextStyle}
              onPress={() => this.setState({ showAnswers: !this.state.showAnswers })}/>
            {this.state.showAnswers && this.renderKidsComments(kids)}
          </View>
        }
      </View>
    )
  }
}
