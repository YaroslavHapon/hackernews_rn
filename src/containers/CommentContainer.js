import { connect } from 'react-redux'
import Comment from './../components/Comment'
import { fetchComment } from './../reducers/comment/get/comment-actions'

const mapStateTopProps = state => ({
  commentGet: state.commentGet
});

const mapDispatchTopProps = dispatch => ({
  fetchComment: (id) => dispatch(fetchComment(id))
});

export default connect(mapStateTopProps, mapDispatchTopProps)(Comment)