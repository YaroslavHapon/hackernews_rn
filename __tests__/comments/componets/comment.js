import React from 'react';
import { ActivityIndicator, View,
  Text, Linking  } from  'react-native';
import { shallow } from 'enzyme';
import sinon from 'sinon'
import Button from './../../../src/components/Button'
import HTMLView from 'react-native-htmlview';
import CommentContainer from './../../../src/containers/CommentContainer'
import Comment from './../../../src/components/Comment'

jest.mock('Linking', () => {
  return {
    openURL: jest.fn()
  }
});

const mockProps = {
  id: 1234,
  commentGet: {},
  fetchComment: sinon.spy(),
  key: 2
};

describe('Testing Comment component', () => {
  const wrapper = shallow(
    <Comment {...mockProps} />
  );

  it('should render correctly', () => {
    wrapper.setProps({ id: 5678 });
    expect(wrapper).toMatchSnapshot();
  });

  it('should call the fetchComment function', () => {
    expect(wrapper.instance().props.fetchComment.calledOnce).toBe(true);
    expect(wrapper.instance().props.fetchComment.args[0][0]).toBe(1234);
  });

  it('should show ActivityIndicator, while request loading', () => {
    const wrapper = shallow(
      <Comment {...mockProps} commentGet= {{ 1234: { loading: true }} } />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(ActivityIndicator).length).toBe(1);
  });

  it('should show ActivityIndicator, when request fails', () => {
    const error = { message: 'Request error!' };
    const wrapper = shallow(
      <Comment {...mockProps} commentGet= {{ 1234: { error: error }}} />
    );
    expect(wrapper.find(View).props().children).toBe(error.message);
  });

  const data = {
    by: 'Jhon Doe',
    text: 'Some text',
    kids: [123, 345, 678]
  };

  it('should render with data', () => {
    const wrapperWithData = shallow(
      <Comment {...mockProps}
        commentGet={ { 1234: { data: data }}}
      />
    );
    expect(wrapperWithData).toMatchSnapshot();
  });

  it('should show the right data', () => {
    const wrapperWithData = shallow(
      <Comment {...mockProps}
        commentGet={ { 1234: { data: data }}}
      />
    );
    expect(wrapperWithData).toMatchSnapshot();
    expect(wrapperWithData.find(Text).nodes[0].props.children[1])
      .toBe(data.by);
  });

  it('should open the comment link in browser', () => {
    const wrapperWithData = shallow(
      <Comment {...mockProps}
        commentGet={ { 1234: { data: data }}}
      />
    );
    const openURlSpy = sinon.spy(Linking, 'openURL');
    wrapperWithData.find(HTMLView).props().onLinkPress();
    expect(openURlSpy.calledOnce).toBe(true);
  });

  it('should show the correct html', () => {
    const wrapperWithData = shallow(
      <Comment {...mockProps}
        commentGet={ { 1234: { data: data }}}
      />
    );
    expect(wrapperWithData.find(HTMLView).props().value)
      .toBe(data.text);
  });

  it('should change the component state', () => {
    const wrapperWithData = shallow(
      <Comment {...mockProps}
        commentGet={{ 1234: { data: data }}}
      />
    );
    wrapperWithData.find(Button).props().onPress();
    expect(wrapperWithData.state().showAnswers).toBe(true);
    expect(wrapperWithData.find(CommentContainer).length).toBe(data.kids.length);
  });
});
