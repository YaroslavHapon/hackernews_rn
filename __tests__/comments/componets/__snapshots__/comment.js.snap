// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Comment component should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingTop": 20,
    }
  }
/>
`;

exports[`Testing Comment component should render with data 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "padding": 10,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontWeight": "bold",
      }
    }
  >
    Author: 
    Jhon Doe
  </Text>
  <HtmlView
    RootComponent={[Function]}
    addLineBreaks={true}
    onError={[Function]}
    onLinkLongPress={null}
    onLinkPress={[Function]}
    value="Some text"
  />
  <View>
    <Button
      buttonTextStyle={
        Object {
          "color": "#333",
          "fontSize": 12,
          "fontWeight": "bold",
          "padding": 5,
        }
      }
      onPress={[Function]}
      style={
        Object {
          "alignItems": "flex-end",
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
      text="Show Answers ↓"
    />
  </View>
</View>
`;

exports[`Testing Comment component should show ActivityIndicator, while request loading 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingTop": 20,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;

exports[`Testing Comment component should show the right data 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "padding": 10,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontWeight": "bold",
      }
    }
  >
    Author: 
    Jhon Doe
  </Text>
  <HtmlView
    RootComponent={[Function]}
    addLineBreaks={true}
    onError={[Function]}
    onLinkLongPress={null}
    onLinkPress={[Function]}
    value="Some text"
  />
  <View>
    <Button
      buttonTextStyle={
        Object {
          "color": "#333",
          "fontSize": 12,
          "fontWeight": "bold",
          "padding": 5,
        }
      }
      onPress={[Function]}
      style={
        Object {
          "alignItems": "flex-end",
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
      text="Show Answers ↓"
    />
  </View>
</View>
`;
